src/Compilers/Core/Portable/InternalUtilities/OneOrMany.cs

```diff
@@ -66,167 +81,133 @@ private bool HasOne
         }
 
         public int Count
-            => HasOne ? 1 : _many.Length;
+            => HasOneItem ? 1 : _many.Length;
 
         public bool IsEmpty
             => Count == 0;
 
-        public OneOrMany<T> Add(T one)
-        {
-            var builder = ArrayBuilder<T>.GetInstance(this.Count + 1);
-            if (HasOne)
-            {
-                builder.Add(_one);
-            }
-            else
-            {
-                builder.AddRange(_many);
-            }
-
-            builder.Add(one);
-            return new OneOrMany<T>(builder.ToImmutableAndFree());
-        }
+        public OneOrMany<T> Add(T item)
+            => HasOneItem ? OneOrMany.Create(_one, item) :
+               IsEmpty ? OneOrMany.Create(item) :
+               OneOrMany.Create(_many.Add(item));
 
         public bool Contains(T item)
-        {
-            if (HasOne)
-                return EqualityComparer<T>.Default.Equals(item, _one);
-
-            foreach (var value in _many)
-            {
-                if (EqualityComparer<T>.Default.Equals(item, value))
-                    return true;
-            }
-
-            return false;
-        }
+            => HasOneItem ? EqualityComparer<T>.Default.Equals(item, _one) : _many.Contains(item);
```

Comment:
don't love it.  but i'm ok with it :)

Reply:
Is there a better way?

Reply:
i'd personally make most methods *not* look at HasOneItem/_one/_many if they don't have to.  do, for example, for `.Contains` i would do:

```
public bool Contains(T item)
{
    foreach (var value in this)
    {
        if (EC<T>.Default.Equals(item, value))
            return true;
    }

    return false;
}
```

(indeed, this could effectively be an extension method).  So we basically just have the core Count/Indexing/Enumeration helpers, and everything else tries to sit on those as much as possible.

--

Note: as above, i don't feel strongly on this.  THis type is so simple that it's also ok for all the methods to just know about hte internal structure and write themselves directly to it.

Reply:
The operations are more efficient when running directly on the `ImmutableArray` though then via a custom enumerator (for >1 element).

You can look at all these methods as two part - specialize for a single element and otherwise delegate to an equivalent method on `ImmutableArray`.

The implementation of some of the methods on `ImmutableArray` are non-trivial so reimplementing them using a custom enumerator would be unnecessarily complicated.

Reply:
totally fair.  efficiency is a fine argument for me to do it the way you did it :)